M2/Macaulay2/e/eigen.cpp

```diff
@@ -0,0 +1,216 @@
+#include <cstdlib>
+
+#if 0
+//void* my_malloc( std::size_t size );
+//void my_free( void* ptr );
+namespace std {
+void* my_malloc( std::size_t size );
+void my_free( void* ptr );
+}
+using std::my_malloc;
+using std::my_free;
+
+#define malloc my_malloc
+#define free my_free
+#endif
+
+#define mpfr foo
+#include "mpreal.h"
+#include <unsupported/Eigen/MPRealSupport>
+#undef mpfr
+#include "eigen.hpp"
+#include <Eigen/SVD>
+
+#if USING_MPIR 
+#include <mpir.h>
+#include <mpirxx.h>
+#else
+#include <gmp.h>
+#include <gmpxx.h>
+#endif
+#include <mpfr.h>
+
+#if 0
+#undef free
+#undef malloc
+
+#include <gc/gc_allocator.h>
+// void* my_malloc( std::size_t size )
+// {
+//   return GC_MALLOC_UNCOLLECTABLE(size);
+// }
+// void my_free( void* ptr )
+// {
+//   GC_FREE(ptr);
+// }
+namespace std {
+void* my_malloc( std::size_t size )
+{
+  return GC_MALLOC_UNCOLLECTABLE(size);
+}
+void my_free( void* ptr )
+{
+  GC_FREE(ptr);
+}
+}
+#endif
+
+/***********************
+ * An attempt to overwrite memory management globally
+ * (i.e., without "#define my_free", etc.)
+ * ... failed!!!
+ * Is there a function that I'm missing???
+  
+void* malloc( std::size_t size )
+{
+  std::cout << "-- std::malloc call!!!" << std::endl;
+  return GC_MALLOC_UNCOLLECTABLE(size);
+}
+void* realloc( void* ptr, std::size_t new_size )
+{
+  std::cout << "-- std::realloc call!!!" << std::endl;
+  GC_FREE(ptr);
+  return GC_MALLOC_UNCOLLECTABLE(new_size);
+}
+void* calloc( std::size_t num, std::size_t size )
+{
+  std::cout << "-- std::calloc call!!!" << std::endl;
+  return GC_MALLOC_UNCOLLECTABLE(num*size);
+}
+void free( void* ptr )
+{
+  std::cout << "-- std::free call!!!" << std::endl;
+  GC_FREE(ptr);
+}
+*/
+
+using Real = foo::mpreal;
```

Comment:
Same here.

Reply:
@mikestillman -- this is for you

to me it looks like the introduction of a synonym for convenience.  Still, why bother?

I don't understand the advantage of "using" over "typedef", or the difference, even.

Reply:
I'll look into this with @antonleykin

`using` plays nicer with templates, and has a more appealing syntax than `typedef`.  In this instance, they are synonyms.